### PR TITLE
Jakarta rewrite infrastructure

### PR DESCRIPTION
**Pull request by gsmet**
Hey @manovotn @Sanne @maxandersen @n1hility @radcortez ,

Here is the promised PR with the Jakarta rewrite infrastructure. It required quite a lot of experiments but I'm quite happy with the result (it looks far more simple in the end than all the experiments I had to do to get to this result). The hardest part was to find a way to apply the changes without ending up with something you can't build anymore as OpenRewrite needs the Maven build to work.

I went until the ArC extension in this first PR. My next step is to be able to build the Hibernate Validator one as it shouldn't be too hard on the HV side but will unblock quite a lot of dependencies as a lot of them are built on the same model and will require the same dependencies for them to work.
It should be far faster now that I figured out a way to orchestrate this thing.

I added a `jakarta/README.md` explaining the general approach and I would be happy to present the approach to you anytime.

Note that this requires (but good news, everything is taken care of in the `jakarta/transform.sh` script so you don't have to do anything related to this):
- a patched version of OpenRewrite (I'm contributing my patches back there so that's until they merged them)
- a patched version of the Rewrite Maven Plugin
- a patched version of the Kotlin Maven Plugin (yes, I know... but I need to be able to skip the main compilation which the Kotlin plugin does not support, it only supports `maven.test.skip` for the test compilation)
- the script also builds SmallRye Config while waiting for the CR2 promised by Roberto

Please see the `jakarta/README.md` before experimenting with this as it has a few advices on how to experiment.

Let's wait for CI on this one as I had to do some more build fixes in addition to the ones that I already merged. These build fixes are in separate commits.

Questions welcome. Happy to present this work in a call.

### CODE REVIEW

File: jakarta/README.md [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+- then sync the branch and run the migration scripts in the separate `quarkus-jakarta` repository
+
+Note that the `transform.sh` script will remove all `999-SNAPSHOT` Quarkus artifacts that are in your local Maven repository.
+
+That way, you can easily reinitialize your `quarkus-jakarta` copy after having attempted a migration.
+
+You can for instance use the following approach:
+
+- create a `quarkus-jakarta` copy of your `quarkus` repository
+- add a `jakarta` remote to this `quarkus-jakarta` copy: `git remote add jakarta ../quarkus/`
```
Comment (Sanne) [inline jakarta/README.md:102]: suggestion:
`git worktree add ../quarkus-jakarta`

File: jakarta/README.md [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+
+- In the [official documentation](https://docs.openrewrite.org/reference/recipes/maven)
+- In the [source for those not already released](https://github.com/gsmet/rewrite/tree/main/rewrite-maven/src/main/java/org/openrewrite/maven) (things are very simple there and well documented in the source)
+
+Recipes are not applied on all POM files, you need to declare the ones you want to apply to a given target POM when changes are necessary.
+Note that in most cases, you don't need to apply any changes.
+
+The active recipes have to be declared in the target POM with a `plugin` block in the `<build><plugins>` section similar to:
+
+```xml
```
Comment (maxandersen) [inline jakarta/README.md:41]: could we not avoid all the pom.xml changes by putting this plugin config in parent and have it pickup `rewrite.yml` ?

**Comment by Sanne** [comment]
It's failing for me with:
```
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  3.377 s
[INFO] Finished at: 2022-03-28T15:01:17+01:00
[INFO] ------------------------------------------------------------------------
[ERROR] No plugin found for prefix 'rewrite' in the current project and in the plugin groups [org.apache.maven.plugins, org.codehaus.mojo] available from the repositories [local (/home/sanne/.m2/repository), jboss-public-repository (https://repository.jboss.org/nexus/content/groups/public/), central (https://repo.maven.apache.org/maven2)] -> [Help 1]
```

Is that a reasonable clue for you?

**Comment by Sanne** [comment]
Attaching the full logs:
[transform.log.gz](https://github.com/quarkusio/quarkus/files/8363648/transform.log.gz)

**Review by Sanne — APPROVED** [review]

**Comment by gsmet** [comment]
OK, so the problem Sanne has is that they released a new version of the rewrite-maven-plugin since I started working on this and it messes things up.
I forked it too to make things more stable.

**Comment by Sanne** [comment]
Excellent, works fine for me as well now.

**Review by radcortez — APPROVED** [review]

**Comment by radcortez** [comment]
Excellent work @gsmet. Thanks!

**Comment by gsmet** [comment]
CI was green. I'm force pushing to squash a !fixup commit.

File: jakarta/README.md (hunk 2 above)
Comment (maxandersen) [reply jakarta/README.md:48]: ack - its just that if there is a rewrite.yml next to the pom.xml my jbang based runner for rewrite could be used in a parallel and removed need to change each pom.xml ...just add rewrite.yml where needed.

**Pull request merged**